src/transformers/models/wav2vec2/tokenization_wav2vec2.py

```diff
@@ -204,31 +224,110 @@ def _convert_id_to_token(self, index: int) -> str:
+                raise ValueError(
+                    "Make sure to provide the `stride` of the audio model "
+                    "in order to retrieve the time stamps for each token."
+                    "The stride is usually defined as `model.stride`."
+                )
+
+            if sampling_rate is None:
+                raise ValueError(
+                    "Make sure to provide the `sampling_rate` of the feature extractor "
+                    "in order to retrieve the time stamps for each token."
+                    "The sampling rate is usually defined as `feature_extractor.sampling_rate`."
+                )
+
+            frame_context_in_seconds = stride / sampling_rate
+
+            end_indices = np.asarray(token_repetitions).cumsum()
+            start_indices = np.concatenate(([0], end_indices[:-1]))
+
+            # create time stamps
+            time_stamps = [
+                {"token": t, "start_time": s * frame_context_in_seconds, "end_time": e * frame_context_in_seconds}
+                for t, s, e in zip(tokens, start_indices, end_indices)
+            ]
+
+            # filter out CTC token
+            time_stamps = list(filter(lambda time_stamp: time_stamp["token"] != self.pad_token, time_stamps))
+
+            if len(time_stamps) != len(processed_tokens):
+                raise ValueError(
+                    f"`time_stamps`: {time_stamps} and `processed_tokens`: {processed_tokens}"
+                    " have to be of the same length, but are: "
+                    f"`len(time_stamps)`: {len(time_stamps)} and `len(processed_tokens)`:"
+                    f" {len(processed_tokens)}"
+                )
+
+            # set tokens to correct
+            for i, token in enumerate(processed_tokens):
+                time_stamps[i]["token"] = token
+
+            # retrieve word time stamps from time stamps
+            word_time_stamps = self._get_word_time_stamps(time_stamps)
+
+        # join to string
+        join_char = " " if spaces_between_special_tokens else ""
+        string = join_char.join(processed_tokens).strip()
 
         if self.do_lower_case:
             string = string.lower()
-        return string
+
+        return {
+            "text": string,
+            "token_time_stamps": time_stamps,
+            "word_time_stamps": word_time_stamps,
+        }
+
+    @staticmethod
+    def _get_word_time_stamps(time_stamps: Dict[str, Union[str, float]]) -> Dict[str, Union[str, float]]:
+        word_time_stamps = []
+        word_begin = time_stamps[0]["token"] != " "
+
+        for i, time_stamp in enumerate(time_stamps):
+            token = time_stamp["token"]
+            word_end = token == " " and i > 0
+            if word_begin:
+                word_time_stamp = {
+                    "word": "",
+                    "start_time": time_stamp["start_time"],
+                }
+
+            if word_end:
+                word_time_stamp["end_time"] = time_stamp["start_time"]
```

Comment:
Should be correct as is - the "end_time" is the start time of the next space. Note that Wav2Vec2 was trained to finish always with a word delimiter boundary

Reply:
Isn't the model trained to output [PAD] on silences ? 

If true, then I think we should exclude [PAD] tokens from the timings. (and so End != start of next)
If [PAD] strictly means "I am no sure" to the model then OK. 


My main concern is that the timings should exclude silences when they exist. Since your code doesn't seem to allow "holes" in the timings I think it cannot account for silences.